data_model/src/events/notification.rs

```diff
@@ -0,0 +1,290 @@
+    pub enum TriggerCompletedOutcome {
+        Success,
+        Failure(String),
+    }
+
+    /// Filter for [`NotificationEvent`]
+    #[derive(
+        Debug,
+        Clone,
+        FromVariant,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Decode,
+        Encode,
+        Deserialize,
+        Serialize,
+        IntoSchema,
+    )]
+    #[ffi_type]
+    #[non_exhaustive]
+    pub enum NotificationEventFilter {
+        AcceptAll,
+        TriggerCompleted(TriggerCompletedEventFilter),
+    }
+
+    /// Filter [`TriggerCompletedEvent`] by
+    /// 1. if `triger_id` is some filter based on trigger id
+    /// 2. if `outcome_type` is some filter based on execution outcome (success/failure)
+    /// 3. if both fields are none accept every event of this type
+    #[derive(
+        Debug,
+        Clone,
+        Constructor,
+        Getters,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Decode,
+        Encode,
+        Deserialize,
+        Serialize,
+        IntoSchema,
+    )]
+    #[ffi_type]
+    #[getset(get = "pub")]
+    #[non_exhaustive]
+    pub struct TriggerCompletedEventFilter {
+        trigger_id: Option<TriggerId>,
+        outcome_type: Option<TriggerCompletedOutcomeType>,
+    }
+
+    /// Type of outcome to filter on
+    #[derive(
+        Debug,
+        Clone,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Decode,
+        Encode,
+        Deserialize,
+        Serialize,
+        IntoSchema,
+    )]
+    #[repr(u8)]
+    #[ffi_type]
+    pub enum TriggerCompletedOutcomeType {
+        Success,
+        Failure,
+    }
+}
+
+#[cfg(feature = "transparent_api")]
+impl super::Filter for NotificationEventFilter {
+    type Event = NotificationEvent;
+
+    /// Check if `self` accepts the `event`.
+    #[inline]
+    fn matches(&self, event: &Self::Event) -> bool {
+        match (self, event) {
+            (Self::AcceptAll, _) => true,
+            (Self::TriggerCompleted(filter), NotificationEvent::TriggerCompleted(event)) => {
+                filter.matches(event)
+            }
+        }
+    }
+}
+
+#[cfg(feature = "transparent_api")]
+impl super::Filter for TriggerCompletedEventFilter {
+    type Event = TriggerCompletedEvent;
+
+    /// Check if `self` accepts the `event`.
+    #[inline]
+    fn matches(&self, event: &Self::Event) -> bool {
+        if let Some(trigger_id) = self.trigger_id() {
+            if trigger_id != event.trigger_id() {
+                return false;
+            }
+        }
+
+        if let Some(outcome_type) = self.outcome_type() {
+            match (outcome_type, event.outcome()) {
+                (TriggerCompletedOutcomeType::Success, TriggerCompletedOutcome::Failure(_))
+                | (TriggerCompletedOutcomeType::Failure, TriggerCompletedOutcome::Success) => {
+                    return false
+                }
+                _ => {}
+            }
+        }
+
+        true
```

Comment:
Isn't it easier to:

1. Remove `return` and use `match` as the last expression
2. Use happy paths inside `match`

?

Reply:
```suggestion
        match (self.outcome_type(), event.outcome()) {
            (None, _) |
            (Some(TriggerCompletedOutcomeType::Success), TriggerCompletedOutcome::Success) |
            (Some(TriggerCompletedOutcomeType::Failure), TriggerCompletedOutcome::Failure(_)) => {
                true
            }
            (Some(_), _ ) => false
        }
```

I mean something like this

Reply:
@Arjentix and then you want to `&&` between to subfilters?

Reply:
After discussion in DM decided to just move `if let Some` into `match`

Reply:
We decided to remove `if let`.